storage/block/block.go

```diff
@@ -31,95 +33,260 @@ import (
 )
 
 var (
-	errReadFromClosedBlock = errors.New("attempt to read from a closed block")
-	errWriteToClosedBlock  = errors.New("attempt to write to a closed block")
+	errReadFromClosedBlock         = errors.New("attempt to read from a closed block")
+	errRetrievableBlockNoRetriever = errors.New("attempt to read from a retrievable block with no retriever")
 
 	timeZero = time.Time{}
 )
 
-// NB(xichen): locking of dbBlock instances is currently done outside the dbBlock struct at the series level.
-// Specifically, read lock is acquired for accessing operations like Stream(), and write lock is acquired
-// for mutating operations like Write(), Reset(), and Close(). Adding a explicit lock to the dbBlock struct might
-// make it more clear w.r.t. how/when we acquire locks, though.
 type dbBlock struct {
-	opts     Options
-	ctx      context.Context
-	start    time.Time
-	segment  ts.Segment
-	checksum uint32
-	closed   bool
+	sync.RWMutex
+
+	opts          Options
+	ctx           context.Context
+	startUnixNano int64
+	segment       ts.Segment
+	length        int
+	checksum      uint32
+
+	accessedUnixNano int64
+
+	mergeTarget DatabaseBlock
+
+	retriever  DatabaseShardBlockRetriever
+	retrieveID ts.ID
+
+	closed bool
 }
 
 // NewDatabaseBlock creates a new DatabaseBlock instance.
 func NewDatabaseBlock(start time.Time, segment ts.Segment, opts Options) DatabaseBlock {
 	b := &dbBlock{
-		opts:   opts,
-		ctx:    opts.ContextPool().Get(),
-		start:  start,
-		closed: false,
+		opts:          opts,
+		ctx:           opts.ContextPool().Get(),
+		startUnixNano: start.UnixNano(),
+		closed:        false,
+	}
+	if segment.Len() > 0 {
+		b.resetSegmentWithLock(segment)
+	}
+	return b
+}
+
+// NewRetrievableDatabaseBlock creates a new retrievable DatabaseBlock instance.
+func NewRetrievableDatabaseBlock(
+	start time.Time,
+	retriever DatabaseShardBlockRetriever,
+	metadata RetrievableBlockMetadata,
+	opts Options,
+) DatabaseBlock {
+	b := &dbBlock{
+		opts:          opts,
+		ctx:           opts.ContextPool().Get(),
+		startUnixNano: start.UnixNano(),
+		closed:        false,
 	}
-	b.resetSegment(segment)
+	b.resetRetrievableWithLock(retriever, metadata)
 	return b
 }
 
+func (b *dbBlock) now() time.Time {
+	nowFn := b.opts.ClockOptions().NowFn()
+	return nowFn()
+}
+
 func (b *dbBlock) StartTime() time.Time {
-	return b.start
+	b.RLock()
+	start := b.startWithLock()
```

Comment:
We could but we don't ever frequently write the start time, hence its just as quick to use an `RLock` here.  Also we extremely rarely call this method, as we iterate over the block start keys of the map the `DatabaseBlock` is in anyway instead of checking the `StartTime() time.Time` for each block.

If there are no writers pending an `RLock` and `RUnlock` is two simple `AddInt32`s anyhow:
```
https://golang.org/src/sync/rwmutex.go?s=1294:1320#L26
```